source/common/network/connection_impl.cc

```diff
@@ -403,6 +430,28 @@ void ConnectionImpl::onFileEvent(uint32_t events) {
     return;
   }
 
+#ifdef __APPLE__
+  // Because OSX doesn't support early close detection via events, we leave the read event enabled
+  // even when reads are disabled. Detect it here and convert it to a Closed event if we can detect
+  // the connection is closed. See https://github.com/envoyproxy/envoy/issues/4294.
+  if (detect_early_close_ && !read_enabled_ && (events & Event::FileReadyType::Read) != 0) {
+    if (detectEarlyClose()) {
+      // No longer connected. Convert to a closed event.
+      ENVOY_CONN_LOG(trace, "early close detection triggered", *this);
+      events |= Event::FileReadyType::Closed;
+    } else {
+      ENVOY_CONN_LOG(trace, "pending read in early close detection", *this);
+      disabled_read_pending_ = true;
```

Comment:
Why isn't this set true under readDisable?  I'm not sure why it would differ from !read_enabled_

Reply:
It indicates that a read event occurred while !read_enabled_. I'll try to think of a better name.